assets/sip-big_vector/big_vector.move

```diff
@@ -0,0 +1,339 @@
+        let result = pop_back(bv);
+        if (i == length(bv)) {
+            result
+        } else {
+            let slice_count = (i / bv.slice_size) + 1;
+            let slice = dynamic_field::borrow_mut<u64, vector<Element>>(&mut bv.id, slice_count);
+            vector::push_back(slice, result);
+            vector::swap_remove(slice, i % bv.slice_size)
+        }
+    }
+
+    /// remove the element at index i and shift the rest elements
+    /// abort when reference more thant 1000 slices
+    /// costly function, use wisely
+    public fun remove<Element: store>(bv: &mut BigVector<Element>, i: u64): Element {
+        let slice = dynamic_field::borrow_mut<u64, vector<Element>>(&mut bv.id, (i / bv.slice_size) + 1);
+        let result = vector::remove(slice, i % bv.slice_size);
+        let slice_count = bv.slice_count;
+        while (slice_count > (i / bv.slice_size) + 1 && slice_count > 1) {
+            let slice = dynamic_field::borrow_mut<u64, vector<Element>>(&mut bv.id, slice_count);
+            let tmp = vector::remove(slice, 0);
+            let prev_slice = dynamic_field::borrow_mut<u64, vector<Element>>(&mut bv.id, slice_count - 1);
+            vector::push_back(prev_slice, tmp);
+            slice_count = slice_count - 1;
+        };
+        trim_slice(bv);
+        bv.length = bv.length - 1;
+
+        result
+    }
+
+    /// drop BigVector, abort if it's not empty
+    public fun destroy_empty<Element: store>(bv: BigVector<Element>) {
+        let BigVector {
+            id,
+            slice_count: _,
+            slice_size: _,
+            length,
+        } = bv;
+        assert!(length == 0, E_NOT_EMPTY);
+        let empty_slice = dynamic_field::remove(&mut id, 1);
+        vector::destroy_empty<Element>(empty_slice);
+        object::delete(id);
+    }
+
+    /// drop BigVector if element has drop ability
+    /// abort when the BigVector contains more thant 1000 slices
+    public fun drop<Element: store + drop>(bv: BigVector<Element>) {
+        let BigVector {
+            id,
+            slice_count,
+            slice_size: _,
+            length: _,
+        } = bv;
+        while (slice_count > 0) {
+            dynamic_field::remove<u64, vector<Element>>(&mut id, slice_count);
+            slice_count = slice_count - 1;
+        };
+        object::delete(id);
+    }
+
+    /// remove empty slice after element removal
+    fun trim_slice<Element: store>(bv: &mut BigVector<Element>) {
+        let slice = dynamic_field::borrow_mut<u64, vector<Element>>(&mut bv.id, bv.slice_count);
+        if (bv.slice_count > 1 && vector::length(slice) == 0) {
+            let empty_slice = dynamic_field::remove(&mut bv.id, bv.slice_count);
+            vector::destroy_empty<Element>(empty_slice);
+            bv.slice_count = bv.slice_count - 1;
+        };
+    }
+}
+
+#[test_only]
+module typus_framework::test_big_vector {
```

Comment:
I think we would want the tests to be a bit more comprehensive for a core data structure with some intricate logic. A few ideas:
- Add a function for checking key internal invariants (`slice_count` is always nonzero, ever slice is `slize_size` except for the last one, `length` is equal to the number of elements, ...), add test-only wrapper functions for all public API's that mutate the data structure, and call this function at the end.
- Shoot for close to 100% test coverage (ok if getters and some abort conditions are not covered, but we would ideally want everything else to be). The current tests are at 84%. https://docs.sui.io/references/cli/move#get-test-coverage-for-a-module